src/gui/notifications/notifier.cpp

```diff
@@ -0,0 +1,97 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "notifier.h"
+
+#include "core/bittorrent/torrenthandle.h"
+#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
+#include "dbusnotifier.h"
+#else
+#include "systemtraynotifier.h"
+#endif
+
+#include <QWidget>
+
+Notifier::Notifier(QObject *parent)
+    : QObject(parent)
+{
+}
+
+namespace
+{
+    inline QString torrentNameOrDefault(const BitTorrent::TorrentHandle* h)
+    {
+        return h ? h->name() : QObject::tr("Torrent");
+    }
+}
+
+void Notifier::notifyTorrentFinished(const BitTorrent::TorrentHandle *h, QWidget *widget)
+{
+    Context ct(widget, h);
+    showNotification(Notifier::TorrentFinished, tr("Download completion"),
+                     tr("%1 has finished downloading.", "e.g: xxx.avi has finished downloading.")
+                        .arg(torrentNameOrDefault(h)), &ct);
+}
+
+void Notifier::notifyTorrentIOError(const BitTorrent::TorrentHandle *h, const QString &message, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::TorrentIOError, tr("I/O Error", "i.e: Input/Output Error"),
+                     tr("An I/O error occurred for torrent %1.\n Reason: %2",
+                        "e.g: An error occurred for torrent xxx.avi.\n Reason: disk is full.")
+                        .arg(torrentNameOrDefault(h)).arg(message), &ct);
+}
+
+void Notifier::notifyUrlDownloadError(const QString &url, const QString &reason, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::UrlDownloadError, tr("Url download error"),
+                     tr("Couldn't download file at url: %1, reason: %2.").arg(url).arg(reason), &ct);
+}
+
+void Notifier::notifySearchFinished(QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::SearchFinished, tr("Search Engine"), tr("Search has finished"), &ct);
+}
+
+void Notifier::showNotification(const QString &title, const QString &message, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::NoType, title, message, &ct);
+}
+
+Notifier *createNotifier(QWidget *mainWindow, QSystemTrayIcon *systrayIcon)
```

Comment:
>I know that you know that I know the difference.

Ok. Sorry.
By the way, I do not think that your run-time example is correct in this case. I would agree with your arguments if we were talking about the problem, the correct solution of which would be so complex and ugly that we could allow some minor violations in favor of ease of implementation. But in this case the correct decision is no less simple than this. Look yourself!
Your:
```c++
Notifier *createNotifier(QWidget *mainWindow, QSystemTrayIcon *systrayIcon)
{
#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
    Q_UNUSED(systrayIcon)
    return new DBusNotifier(mainWindow, mainWindow);
#else
    return new SystemTrayNotifier(mainWindow, systrayIcon);
#endif
}
```
Correct (in mainwindow.cpp):
```c++
#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
    notifier = new DBusNotifier(mainWindow, mainWindow);
#else
    notifier =  new SystemTrayNotifier(mainWindow, systrayIcon);
#endif
```
The incorrectness of your implementation, you can prove by applying a semblance method of mathematical induction. Suppose we have implemented another notifier (SomeOtherNotifier), which requires an instance of some widget (SomeWidget) and doesn't require MainWindow instance. Then we get the following:
Your:
```c++
Notifier *createNotifier(QWidget *mainWindow, QSystemTrayIcon *systrayIcon, SomeWidget *someWidget)
{
#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
    Q_UNUSED(systrayIcon)
    Q_UNUSED(someWidget)
    return new DBusNotifier(mainWindow, mainWindow);
#else if (some_condition)
    Q_UNUSED(mainWindow)
    Q_UNUSED(systrayIcon)
    return new SomeOtherNotifier(someWidget);
#else
    Q_UNUSED(someWidget)
    return new SystemTrayNotifier(mainWindow, systrayIcon);
#endif
}
```
Correct (in mainwindow.cpp):
```c++
#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
    notifier = new DBusNotifier(mainWindow, mainWindow);
#else if (some_condition)
    notifier = new SomeOtherNotifier(someWidget);
#else
    notifier = new SystemTrayNotifier(mainWindow, systrayIcon);
#endif
```
And so on...